core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -103,6 +103,17 @@ abstract class RDD[T: ClassTag](
     _sc
   }
 
+  /**
+   * Lock for all mutable state of this RDD (persistence, partitions, dependencies, etc.).  We do
+   * not use `this` because RDDs are user-visible, so users might have added their own locking on
+   * RDDs; sharing that could lead to a deadlock.
+   *
+   * One thread might hold the lock on many of these, for a chain of RDD dependencies; but
+   * because DAGs are acyclic, and we only ever hold locks for one path in that DAG, there is no
+   * chance of deadlock.
+   */
+  private val stateLock = new SerializableObject()
```

Comment:
Does `@transient` work here? or will it not be reinitialized when the object is deserialized? `@transient lazy` probably works but may be overkill. Rather than define a class (which isn't a big deal), just use some random serializable object like `new Integer(0)`?

Reply:
`@transient` doesn't work because somebody *could* call `dependencies` or `partitions` in their map functions, and then it would fail trying to get a lock on null.  (weird use case, but I wouldn't break compatibility over it.)  Sure I could switch to `Integer(0)` if you prefer -- the purpose seems less clear to me, but I can explain in a comment.

Reply:
Nah, neutral on it. Up to you.

Reply:
I went ahead and made the change.  I also added a test case for this.

Honestly, the test case is pretty convoluted ... it kinda makes me wonder if its even worth having a test case, or just making it `@transient` after all and breaking compatibility for this very unusual use.  I can imagine having some slightly more "normal" uses in a real case (eg. maybe changing the processing based on the number of partitions in an RDD), but nothing really compelling.

Reply:
What's the failure mode here? an RDD shouldn't be accessed anywhere but the driver, so if it's serialized somewhere else and accessed, that's a problem already, no? or is it more complex than that?

Reply:
Actually RDDs are serialized as part of computing them on executors.

But I'm not sure if there are well-defined semantics about what is safe to access in executors. For example, if your custom RDD tries to access `dependencies` on the executor side, is it guaranteed that it will be available? Because actually calculating the dependencies seems like something only the driver can do.

 (`partitions_` is `@transient` so it seems clear that it's not to be used in executors).

My gut feeling is that all of these things should be transient; it feels that `RDD.compute` really should be independent of any (Spark-internal) state in the RDD class. But it's hard to say if that's really safe to enforce at this point, with so many years of people writing their own RDDs.

Reply:
Yeah I recall the RDDs get serialized as a side effect, but that is mostly accidental? you can't perform operations on it. At least, I'd say that one shouldn't assume anything about an RDD works once it's on the executors. I doubt that just making it transient (thus null on the executors) breaks anything that happened to work now?

Reply:
You're talking about an app using RDDs. I'm talking about someone *implementing* an RDD. The only way that works is through serialization, because `compute` is a method of the RDD class. So it's not accidental, it's by design.

Reply:
OK, got it. But does it change much? would a custom RDD reasonably guarantee some semantics as used on executor? I'd assume all bets are off.

Reply:
It's not about what custom RDDs guarantee, it's about what Spark does.

If previously `dependencies` was available on the executor and now it's not, it might break things if someone depended on that.

You could make the lock transient and check whether it is null and avoid the locking on the executor side, but that seems like a marginal gain (few bytes of serialized data).

Reply:
I agree with both of you, and I'm torn myself -- its a super weird thing to do, but the fact is you could do it before.  Should I break it now?  I decided not to, since it wasn't too hard to keep it working.

> (`partitions_` is `@transient` so it seems clear that it's not to be used in executors).

I dunno if that is true -- sure that `private var` is transient but the public method you actually call, `partitions`, it'll check for null and regenerate the value.

... except, that it turns out that it doesn't work for some built in RDDs, like ParallelCollectionRDD, because other data it needs are also `transient`.  But a user could have their own custom RDD which it works fine for.  I started to write up a test case like this but decided the test was getting ridiculously involved for a "feature" we're not even sure we want to support.

> You could make the lock transient and check whether it is null and avoid the locking on the executor side, but that seems like a marginal gain (few bytes of serialized data).

doesn't sounds like its worth the added complexity in the code to me.